File: ejemplos/nerve/06-puente-de-progreso-gif/ejemplo.go
/*
 * Ejemplo: Dibujar una barra de progreso nativa en Go
 * 
 * Qué enseña este ejemplo:
 *   - Cómo usar la librería nativa de Go `image` y `image/draw`.
 *   - Cómo dibujar rectángulos en un lienzo RGBA.
 *   - Cómo codificar y guardar una imagen en formato PNG (image/png).
 * 
 * Para tu reto (Puente de progreso GIF con Nerve):
 *   En lugar de un porcentaje fijo, tu receptor de Nerve recibirá actualizaciones
 *   (ej. 10%, 20%), y con cada actualización dibujarás la barra correspondiente 
 *   y guardarás la imagen.
 * 
 * Glosario:
 *   image.NewRGBA()       — Crea una nueva imagen de un tamaño específico que soporta color.
 *   color.RGBA{}          — Define un color usando rojo, verde, azul y alpha (transparencia).
 *   draw.Draw()           — Copia una imagen/color sobre otra en un área específica.
 *   png.Encode()          — Escribe los datos de la imagen en formato PNG a un archivo.
 */

package main

import (
	"fmt"
	"image"
	"image/color"
	"image/draw"
	"image/png"
	"log"
	"os"
)

func generarBarraProgreso(porcentaje int, rutaSalida string) {
	ancho := 400
	alto := 50

	// Aseguramos que el porcentaje esté entre 0 y 100
	if porcentaje < 0 {
		porcentaje = 0
	} else if porcentaje > 100 {
		porcentaje = 100
	}

	// Calculamos el ancho de la barra
	anchoProgreso := int((float64(porcentaje) / 100.0) * float64(ancho))

	// 1. Definimos los colores
	colorFondo := color.RGBA{220, 220, 220, 255} // Gris claro
	colorBarra := color.RGBA{50, 205, 50, 255}   // Verde lima

	// 2. Creamos el lienzo (la imagen base)
	// image.Rect(x0, y0, x1, y1) define los límites
	lienzo := image.NewRGBA(image.Rect(0, 0, ancho, alto))

	// 3. Llenamos todo el lienzo con el color de fondo
	// draw.Src significa que reemplazará el pixel completamente
	draw.Draw(lienzo, lienzo.Bounds(), &image.Uniform{colorFondo}, image.Point{}, draw.Src)

	// 4. Dibujamos el rectángulo de progreso
	// Creamos un rectángulo que va desde (0,0) hasta (anchoProgreso, alto)
	rectProgreso := image.Rect(0, 0, anchoProgreso, alto)
	draw.Draw(lienzo, rectProgreso, &image.Uniform{colorBarra}, image.Point{}, draw.Src)

	// 5. Guardamos la imagen en un archivo
	archivo, err := os.Create(rutaSalida)
	if err != nil {
		log.Printf("Error al crear el archivo %s: %v", rutaSalida, err)
		return
	}
	defer archivo.Close()

	if err := png.Encode(archivo, lienzo); err != nil {
		log.Printf("Error al codificar PNG: %v", err)
		return
	}

	fmt.Printf("✅ Imagen guardada en: %s con un progreso de %d%%\n", rutaSalida, porcentaje)
}

func main() {
	fmt.Println("Generando ejemplos de barras de progreso...\n")

	generarBarraProgreso(25, "progreso_25.png")
	generarBarraProgreso(75, "progreso_75.png")
	generarBarraProgreso(100, "progreso_100.png")

	fmt.Println("\nRevisa tu carpeta, deberías ver tres archivos .png nuevos.")
}

File: ejemplos/nerve/06-puente-de-progreso-gif/ejemplo.py
"""
Ejemplo: Dibujar una barra de progreso con Pillow (Python)

Qué enseña este ejemplo:
  - Cómo crear una imagen desde cero.
  - Cómo dibujar formas geométricas (rectángulos) sobre la imagen.
  - Cómo guardar la imagen generada en el disco.

Para tu reto (Puente de progreso GIF con Nerve):
  En lugar de un porcentaje fijo, tu receptor de Nerve recibirá actualizaciones
  (ej. 10%, 20%), y con cada actualización dibujarás la barra correspondiente 
  y guardarás la imagen para que una página web u otro cliente pueda verla.

Glosario:
  Image.new()      — Crea un lienzo en blanco. Pide el modo ("RGB"), tamaño (ancho, alto) y color de fondo.
  ImageDraw.Draw() — Crea un objeto que permite dibujar sobre la imagen.
  draw.rectangle() — Dibuja un rectángulo dadas las coordenadas [x0, y0, x1, y1] y un color.
"""

from PIL import Image, ImageDraw

def generar_barra_progreso(porcentaje: int, ruta_salida: str):
    """
    Crea una imagen con una barra de progreso que representa el porcentaje dado.
    """
    # 1. Definimos dimensiones de la imagen
    ancho = 400
    alto = 50

    # 2. Creamos una nueva imagen con fondo gris claro
    color_fondo = (220, 220, 220)  # RGB
    imagen = Image.new("RGB", (ancho, alto), color_fondo)

    # 3. Preparamos el objeto para dibujar
    draw = ImageDraw.Draw(imagen)

    # 4. Calculamos el ancho de la barra verde (progreso)
    # Si porcentaje es 50, el ancho será el 50% de 400 = 200.
    porcentaje = max(0, min(100, porcentaje)) # Aseguramos que esté entre 0 y 100
    ancho_progreso = int((porcentaje / 100) * ancho)

    # 5. Dibujamos el rectángulo verde
    # Las coordenadas son [x_inicial, y_inicial, x_final, y_final]
    # Empieza en (0,0) y termina en (ancho_progreso, alto)
    color_barra = (50, 205, 50)  # Verde lima
    draw.rectangle([0, 0, ancho_progreso, alto], fill=color_barra)

    # 6. Guardamos la imagen
    imagen.save(ruta_salida)
    print(f"✅ Imagen guardada en: {ruta_salida} con un progreso de {porcentaje}%")

def main():
    print("Generando ejemplos de barras de progreso...\n")
    
    # Generamos tres imágenes distintas
    generar_barra_progreso(25, "progreso_25.png")
    generar_barra_progreso(75, "progreso_75.png")
    generar_barra_progreso(100, "progreso_100.png")
    
    print("\nRevisa tu carpeta, deberías ver tres archivos .png nuevos.")

if __name__ == "__main__":
    main()
